### PR TITLE
Adding static analysis tool with dependency checker rules

### PR DESCRIPTION
**Pull request by markcowl**

### CODE REVIEW

File: tools/StaticAnalysis/ConsoleLogger.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    /// Simple class for logging information
+    /// </summary>
+    public class ConsoleLogger : AnalysisLogger
+    {
+
+        public ConsoleLogger(string baseDirectory) : base(baseDirectory)
+        {
+        }
+
+       public override void WriteError(string error)
```
Comment (stankovski) [inline tools/StaticAnalysis/ConsoleLogger.cs]: Indentation

File: tools/StaticAnalysis/ConsoleLogger.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            Console.WriteLine(string.Format("### ERROR {0}", error));
```
Comment (stankovski) [inline tools/StaticAnalysis/ConsoleLogger.cs]: string.Format is redundant: `Console.WriteLine("### ERROR {0}", error);` is sufficient. 

Same on line 43.

File: tools/StaticAnalysis/Decorator.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            _action(record);
+            if (Inner != null)
+            {
+                Inner.Apply(record);
+            }
+        }
+
+        public void AddDecorator(Action<T> action, string name)
+        {
+            if (Inner == null)
```
Comment (stankovski) [inline tools/StaticAnalysis/Decorator.cs]: I would add a null check for parameters either here or in ctor

File: tools/StaticAnalysis/DependencyAnalyzer/AssemblyMetadata.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    [Serializable]
+    public class AssemblyMetadata
+    {
+        AssemblyName _name;
+        string _location;
+        IList<AssemblyName> _references;
+
+        public AssemblyMetadata(Assembly assembly)
+        {
+            _name = assembly.GetName();
```
Comment (stankovski) [inline tools/StaticAnalysis/DependencyAnalyzer/AssemblyMetadata.cs]: null check for assembly

File: tools/StaticAnalysis/DependencyAnalyzer/AssemblyNameComparer.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace StaticAnalysis.DependencyAnalyzer
+{
+    public class AssemblyNameComparer : IEqualityComparer<AssemblyName>
+    {
+        public static string GetComparisonName(AssemblyName name)
+        {
+            return string.Format("{0}.v{1}", name.Name, name.Version);
```
Comment (stankovski) [inline tools/StaticAnalysis/DependencyAnalyzer/AssemblyNameComparer.cs]: null check for name

File: tools/StaticAnalysis/DependencyAnalyzer/AssemblyRecord.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+
+namespace StaticAnalysis.DependencyAnalyzer
+{
+    public class AssemblyEqualityComparer : IEqualityComparer<AssemblyRecord>
```
Comment (stankovski) [inline tools/StaticAnalysis/DependencyAnalyzer/AssemblyRecord.cs]: This doesn't seem to be used anywhere...

File: tools/StaticAnalysis/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Comment (stankovski) [inline tools/StaticAnalysis/Properties/AssemblyInfo.cs]: Header

File: tools/StaticAnalysis/IReportRecord.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
```
Comment (stankovski) [inline tools/StaticAnalysis/IReportRecord.cs]: Header

File: tools/StaticAnalysis/DependencyAnalyzer/SharedAssemblyConflict.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
+﻿using System;
```
Comment (stankovski) [inline tools/StaticAnalysis/DependencyAnalyzer/SharedAssemblyConflict.cs]: Header

File: tools/StaticAnalysis/Program.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            var directories = new List<string>
+            {
+                Path.Combine(installDir, @"ResourceManager\AzureResourceManager\"),
+                Path.Combine(installDir, @"ServiceManagement\Azure\")
+            };
+
+            var reportsDirectory = Directory.GetCurrentDirectory();
+            if (args.Length > 1 && Directory.Exists(args[1]))
+            {
+                reportsDirectory = args[1];
```
Comment (hovsepm) [inline tools/StaticAnalysis/Program.cs]: Please add warning message that will indicate parameter 2 is not a valid folder and the output will be done in the current folder (and print that out so we would not wonder where was the current folder on the CI machine)

File: tools/StaticAnalysis/DependencyAnalyzer/DependencyAnalyzer.cs [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+                    Logger.WriteMessage("Processing Directory {0}", directoryPath);
+                    _assemblies.Clear();
+                    _versionConflictLogger.Decorator.AddDecorator(r => { r.Directory = directoryPath; } , "Directory");
+                     _missingAssemblyLogger.Decorator.AddDecorator(r => { r.Directory = directoryPath; } , "Directory");
+                      _extraAssemblyLogger.Decorator.AddDecorator(r => { r.Directory = directoryPath; } , "Directory");
+                  ProcessDirectory(directoryPath);
+                    _versionConflictLogger.Decorator.Remove("Directory");
+                     _missingAssemblyLogger.Decorator.Remove("Directory");
+                    _extraAssemblyLogger.Decorator.Remove("Directory");
+               }
```
Comment (hovsepm) [inline tools/StaticAnalysis/DependencyAnalyzer/DependencyAnalyzer.cs]: indentation seems off from line 65 to 71.

File: tools/StaticAnalysis/Decorator.cs (hunk 3 above)
Comment (markcowl) [inline tools/StaticAnalysis/Decorator.cs]: Yeah, I think this is a good practice in general

**Pull request merged**